sdks/java/extensions/smb/src/main/java/org/apache/beam/sdk/extensions/smb/SortedBucketSink.java

```diff
@@ -0,0 +1,359 @@
+import org.apache.beam.sdk.values.PInput;
+import org.apache.beam.sdk.values.POutput;
+import org.apache.beam.sdk.values.PValue;
+import org.apache.beam.sdk.values.TupleTag;
+import org.apache.beam.vendor.guava.v20_0.com.google.common.base.Supplier;
+import org.apache.beam.vendor.guava.v20_0.com.google.common.collect.ImmutableList;
+import org.apache.beam.vendor.guava.v20_0.com.google.common.collect.ImmutableMap;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+/**
+ * Writes a PCollection representing sorted, bucketized data to files, where the # of files is equal
+ * to the # of buckets, assuming the contents of each bucket fit on a single worker.
+ *
+ * @param <SortingKeyT>
+ * @param <ValueT>
+ */
+public class SortedBucketSink<SortingKeyT, ValueT>
+    extends PTransform<PCollection<ValueT>, WriteResult> {
+
+  private final BucketMetadata<SortingKeyT, ValueT> bucketingMetadata;
+  private final SMBFilenamePolicy smbFilenamePolicy;
+  private final Supplier<Writer<ValueT>> writerSupplier;
+  private final ResourceId tempDirectory;
+  private final Coder<SortingKeyT> sortingKeyCoder;
+
+  public SortedBucketSink(
+      BucketMetadata<SortingKeyT, ValueT> bucketingMetadata,
+      Coder<SortingKeyT> sortingKeyCoder,
+      SMBFilenamePolicy smbFilenamePolicy,
+      Supplier<Writer<ValueT>> writerSupplier,
+      ResourceId tempDirectory) {
+    this.bucketingMetadata = bucketingMetadata;
+    this.sortingKeyCoder = sortingKeyCoder;
+    this.smbFilenamePolicy = smbFilenamePolicy;
+    this.writerSupplier = writerSupplier;
+    this.tempDirectory = tempDirectory;
+  }
+
+  @Override
+  public final WriteResult expand(PCollection<ValueT> input) {
+    final Coder<KV<Integer, KV<SortingKeyT, ValueT>>> bucketedCoder =
+        KvCoder.of(VarIntCoder.of(), KvCoder.of(sortingKeyCoder, input.getCoder()));
+
+    return input
+        .apply(
+            "Assign buckets",
+            ParDo.of(new ExtractBucketAndSortKey<>(this.bucketingMetadata, this.sortingKeyCoder)))
+        .setCoder(bucketedCoder)
+        .apply("Group per bucket", GroupByKey.create())
+        .apply("Sort values in bucket", SortValues.create(BufferedExternalSorter.options()))
+        .apply(
+            "Write bucket data",
+            new WriteOperation<>(
+                smbFilenamePolicy, bucketingMetadata, writerSupplier, tempDirectory));
+  }
+
+  /*
+   *
+   */
+  static final class ExtractBucketAndSortKey<K, V> extends DoFn<V, KV<Integer, KV<K, V>>> {
+    private final BucketMetadata<K, V> bucketMetadata;
+    private final Coder<K> sortingKeyCoder;
+
+    ExtractBucketAndSortKey(BucketMetadata<K, V> bucketMetadata, Coder<K> sortingKeyCoder) {
+      this.bucketMetadata = bucketMetadata;
+      this.sortingKeyCoder = sortingKeyCoder;
+    }
+
+    @ProcessElement
+    public void processElement(ProcessContext c) throws Exception {
+      final V record = c.element();
+      final K key = bucketMetadata.extractSortingKey(record);
```

Comment:
If `BucketMetadata` can infer coder from `keyClass`, this whole block can be moved there as something like `BucketMetadata#buildBucket(record): KV[Int, KV[Array[Byte], V]]`?

Also if we require `K` has a deterministic coder (fairly reasonable, Beam does that), we can replace `K` with `Array[Byte]` in the output type and sort by serialized bytes.

(`Array[Byte]` can also be protobuf `ByteString` if that makes comparison easier)

Reply:
hm, I moved this function to the `SortedBucketSink` class originally because I wanted it to have access to the `Pipeline`'s coder registry--`BucketMetadata` can't necessarily infer a coder without that. Seems like each `BucketMetadata` impl could provide a map of Coder overrides for different key types (i.e. a `CharSequence` coder for Avro metadata), and if there's no override found for our key class, then we fall back to whatever's in the `Pipeline`'s coder registry.

Reply:
Yeah I think `BucketMetadata` doesn't need access to the user defined `CoderRegistry` since the key is most likely a primitive type. `CoderRegistry.createDefault().getCoder(keyClass)` should be enough? `AvroBucketMetadata` can start with a default registry and add ones for `CharSequence` and `ByteBuffer`.